lock/src/seqlock.rs

```diff
@@ -170,6 +177,30 @@ impl<'s, T> ReadGuard<'s, T> {
         mem::forget(self);
         result
     }
+
+    pub fn upgrade(self) -> Result<WriteGuard<'s, T>, ()> {
```

Comment:
I suggest we implement the upgrade functionality in `RawSeqLock` and use it in `SeqLock`.